calicoctl/commands/migrateipam.go

```diff
@@ -0,0 +1,301 @@
+	Revision string
+	TTL      time.Duration // For writes, if non-zero, key has a TTL.
+}
+
+// ipamResults contains the results from executing an IPAM backend command
+// TODO: May not need this if we decide not to print out the resources that were imported
+type ipamResults struct {
+	// The number of resources that are being configured.
+	numResources int
+
+	// The number of resources that were actually configured.  This will
+	// never be 0 without an associated error.
+	numHandled int
+
+	// The results returned from each invocation
+	resources []*model.KVPair
+
+	// Errors associated with individual resources
+	resErrs []error
+}
+
+func NewMigrateIPAM(c client.Interface) *migrateIPAM {
+	type accessor interface {
+		Backend() bapi.Client
+	}
+	bc := c.(accessor).Backend()
+	return &migrateIPAM{
+		client: bc,
+	}
+}
+
+func (m *migrateIPAM) PullFromDatastore() error {
+	ctx := context.Background()
+
+	blockKVList, err := m.client.List(ctx, model.BlockListOptions{}, "")
+	if err != nil {
+		return err
+	}
+
+	blockAffinityKVList, err := m.client.List(ctx, model.BlockAffinityListOptions{}, "")
+	if err != nil {
+		return err
+	}
+
+	ipamHandleKVList, err := m.client.List(ctx, model.IPAMHandleListOptions{}, "")
+	if err != nil {
+		return err
+	}
+
+	ipamConfigKV, err := m.client.Get(ctx, model.IPAMConfigKey{}, "")
+	if err != nil {
+		// If the resource does not exist, do not throw the error
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			return err
+		}
+	}
+
+	// Convert all of the abstract KV Pairs into the appropriate types.
+	blocks := []*IPAMBlockKVPair{}
+	for _, item := range blockKVList.KVPairs {
+		modelBlockKey, ok := item.Key.(model.BlockKey)
+		if !ok {
+			return fmt.Errorf("Could not convert %+v to a BlockKey", item.Key)
+		}
+
+		// Convert this to a key that has values for json encoding.
+		blockKey := BlockKey{
+			CIDR: modelBlockKey.CIDR,
+		}
+
+		block, ok := item.Value.(*model.AllocationBlock)
+		if !ok {
+			return fmt.Errorf("Could not convert %+v to an AllocationBlock", item.Value)
+		}
+		blocks = append(blocks, &IPAMBlockKVPair{
+			Key:      &blockKey,
+			Value:    block,
+			Revision: item.Revision,
+			TTL:      item.TTL,
+		})
+	}
+
+	blockAffinities := []*BlockAffinityKVPair{}
+	for _, item := range blockAffinityKVList.KVPairs {
+		modelBlockAffinityKey, ok := item.Key.(model.BlockAffinityKey)
+		if !ok {
+			return fmt.Errorf("Could not convert %+v to a BlockAffinityKey", item.Key)
+		}
+
+		// Convert this to a key that has vlaues for json encoding.
+		blockAffinityKey := BlockAffinityKey{
+			CIDR: modelBlockAffinityKey.CIDR,
+			Host: modelBlockAffinityKey.Host,
+		}
+
+		blockAffinity, ok := item.Value.(*model.BlockAffinity)
+		if !ok {
+			return fmt.Errorf("Could not convert %+v to a BlockAffinity", item.Value)
+		}
+		blockAffinities = append(blockAffinities, &BlockAffinityKVPair{
+			Key:      &blockAffinityKey,
+			Value:    blockAffinity,
+			Revision: item.Revision,
```

Comment:
Do we care about revision and TTL information for the purposes of migration? I think probably not?

Reply:
Revision information we definitely don't need. Not quite sure about TTL? Not sure if there was a value in the old datastore if we would want it moved over or since it's a new datastore, we don't care about it.